lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2551,139 +2553,174 @@ buildAndSignTransactionNew
        , GenChange s
        )
     => Tracer IO WalletLog
-    -> TimeInterpreter (Either PastHorizonException)
+    -> TimeInterpreter (ExceptT PastHorizonException IO)
     -> DBLayer IO s k
     -> NetworkLayer IO Block
     -> TransactionLayer k ktype SealedTx
+    -> Passphrase "user"
     -> WalletId
     -> ArgGenChange s
     -> AnyRecentEra
-    -> Passphrase "user"
     -> PreSelection
     -> TransactionCtx
     -> IO (Tx, TxMeta, UTCTime, SealedTx)
-buildAndSignTransactionNew
-    tr ti db netLayer txLayer walletId changeGen era pwd preSelection txCtx =
+buildAndSignTransactionNew tr ti db netLayer txLayer pwd walletId
+    argGenChange era preSelection txCtx = do
+    --
+    utxo@(_, wallet, _) <- throwOnError $
+        withExceptT (ExceptionConstructTx . ErrConstructTxNoSuchWallet) $
+            readWalletUTxOIndex @_ @_ @k db walletId
+
+    pureTimeInterpreter <- snapshot ti
+    protocolParams <- currentProtocolParameters netLayer
+
+    (tx, meta, sealed) <- throwOnError $
+        let wrapError = ExceptionWitnessTx . ErrWitnessTxWithRootKey
+         in withRootKey db walletId pwd wrapError $ \rootKey scheme ->
+                withExceptT (either ExceptionBalanceTx ExceptionConstructTx) $
+                    buildAndSignTransactionPure @k @ktype @s @n
+                        tr
+                        pureTimeInterpreter
+                        utxo
+                        rootKey
+                        scheme
+                        pwd
+                        protocolParams
+                        txLayer
+                        argGenChange
+                        era
+                        preSelection
+                        txCtx
+
+    tipTime <- interpretQuery
+        (neverFails "buildAndSignTransactionNew: slot is ahead of the node tip"
+            (hoistTimeInterpreter except pureTimeInterpreter))
+        (slotToUTCTime (currentTip wallet ^. #slotNo))
+    pure (tx, meta, tipTime, sealed)
+  where
+    throwOnError = either throwIO pure <=< runExceptT
+
+buildAndSignTransactionPure
+    :: forall k ktype s (n :: NetworkDiscriminant) m
+     . ( Typeable n
+       , Typeable s
+       , Typeable k
+       , WalletKey k
+       , HardDerivation k
+       , BoundedAddressLength k
+       , Bounded (Index (AddressIndexDerivationType k) (AddressCredential k))
+       , IsOwned s k ktype
+       , IsOurs s RewardAccount
+       , GenChange s
+       , MonadRandom m
+       )
+    => Tracer m WalletLog
+    -> TimeInterpreter (Either PastHorizonException)
+    -> (UTxOIndex WalletUTxO, Wallet s, Set Tx)
+    -> k 'RootK XPrv
+    -> PassphraseScheme
+    -> Passphrase "user"
+    -> ProtocolParameters
+    -> TransactionLayer k ktype SealedTx
+    -> ArgGenChange s
+    -> AnyRecentEra
+    -> PreSelection
+    -> TransactionCtx
+    -> ExceptT (Either ErrBalanceTx ErrConstructTx) m (Tx, TxMeta, SealedTx)
```

Comment:
Might we not as well drop the `m` and return `Either`? Otherwise the caller who intends to use this with `modifyDBMaybe` will need to figure out how to run the `m`.

Reply:
`balanceTransaction` has type with `m` already `ExceptT ErrBalanceTx m (Cardano.Tx era)`